libs/polysemy-wire-zoo/test/Test/IntersperseSpec.hs

```diff
@@ -27,6 +29,25 @@ spec = do
               ]
     result `shouldBe` desired
 
+  -- Example showing how intersperse lays out actions
+  it "should stick code before every action" $ do
+    let result =
+          fst $ run $
+            runTraceList $
+              outputToTrace show $
+                evalState @Int 0 $
+                  intersperse ((output =<< get) <* modify (+1)) $ do
+                    -- 0
+                    trace "start"
+                    pure ()
+                    -- 1
+                    trace "middle"
+                    -- 2
+                    _ <- get
+                    -- 3
+                    trace "end"
+    result `shouldBe` [ "0", "start", "1", "middle", "2", "3", "end" ]
```

Comment:
@isovector , I would expect the result to be `["0", "start", "1", "2", "middle", "3", "4", "end"]`, i.e., to another `modify (+1)` to happen between `trace "start"` and `pure ()`. Why is that a wrong expectation? Is it because `pure ()` is not a binding operation?

Reply:
By the monad laws, `m1 >>= \a -> pure () >> m2 a = m1 >>= m2`, so you shouldn't expect the existence of a `pure ()` in the middle of a do block to be detectable.

Reply:
Right! Thanks for that.

Reply:
Furthermore, the combinator works by rewriting calls to `send`, which the standard `m`-quantified `Monad m => ...` operations do not use.